File: Python/currency_converter.py
import urllib.request
import json

API_URL = "https://api.fxratesapi.com/latest?api_key=fxr_live_ba7a79734cfdde9921de59c64463466d3070"

def fetch_exchange_rates():
    try:
        with urllib.request.urlopen(API_URL) as response:
            data = response.read()
            exchange_rates = json.loads(data)
            if exchange_rates.get("success"):
                return exchange_rates
            else:
                print("Error fetching data.")  #exchange_rates.get("error", "Unknown error")
                return None
    except Exception as e:
        print("An error occurred:", e)
        return None

def convert_currency(rates, myBase, target, amount):
    if myBase not in rates and myBase != "USD":
        print(f"Currency {myBase} not found.")
        return None     #so to not go to next if
    if target not in rates:
        print(f"Currency {target} not found.")
        return None

    if myBase != "USD":
        amount_in_usd = amount / rates[myBase]
    else:
        amount_in_usd = amount

    converted_amount = amount_in_usd * rates[target]
    return converted_amount

def main():
    data = fetch_exchange_rates()
    if not data:
        return

    rates = data.get("rates", {})
    base_currency = data.get("base", "USD")
    rates[base_currency] = 1  # Ensure the base currency itself is in the rates list

    print(f"Exchange rates base currency: {base_currency}")
    print("Available currencies:", ", ".join(rates.keys()))

    while True:
        print("\nCurrency Converter")
        from_currency = input("Enter the source currency (e.g., USD): ").upper()
        if from_currency not in rates and from_currency != base_currency:
            print(f"{from_currency} is not available.")
            continue

        to_currency = input("Enter the target currency (e.g., EUR): ").upper()
        if to_currency not in rates:
            print(f"{to_currency} is not available.")
            continue

        try:
            amount = float(input("Enter the amount to convert: "))
        except ValueError:
            print("Invalid amount. Please enter a numeric value.")
            continue

        result = convert_currency(rates, from_currency, to_currency, amount)
        if result is not None:
            print(f"{amount} {from_currency} is approximately {result:.3f} {to_currency}")

        reverse = input("Would you like to reverse the conversion? (yes/no): ").strip().lower()
        if reverse == "yes":
            reverse_result = convert_currency(rates, to_currency, from_currency, amount)
            if reverse_result is not None:
                print(f"{amount} {to_currency} is equivalent to {reverse_result:.2f} {from_currency}\n")
        
        again = input("Do you want to convert again, with other currencies? (yes/no): ").lower()
        if again != "yes":
            break

if __name__ == "__main__":
    main()

File: CSharp/CurrencyConverter.cs
using System;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using System.Collections.Generic;

namespace CurrencyConverterApp
{
    class CurrencyConverter
    {
        static async Task Main(string[] args)
        {
            Console.WriteLine("Currency Converter");

            Console.Write("Enter base currency (e.g., USD): ");
            string baseCurrency = Console.ReadLine().ToUpper();

            Console.Write("Enter target currency (e.g., EUR): ");
            string targetCurrency = Console.ReadLine().ToUpper();

            Console.Write("Enter amount: ");
            if (!double.TryParse(Console.ReadLine(), out double amount))
            {
                Console.WriteLine("Invalid amount. Please enter a numeric value.");
                return;
            }

            try
            {
                double convertedAmount = await ConvertCurrency(baseCurrency, targetCurrency, amount);
                Console.WriteLine($"{amount} {baseCurrency} = {convertedAmount:F2} {targetCurrency}");
                Console.WriteLine("Would you like to reverse the conversion? (y/n)");
                string reverse = Console.ReadLine().ToLower();
                if (reverse == "y") {
                    convertedAmount = await ConvertCurrency(targetCurrency, baseCurrency, amount);
                    Console.WriteLine($"{amount} {targetCurrency} = {convertedAmount:F2} {baseCurrency}");
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error: {ex.Message}");
            }
        }

        static async Task<double> ConvertCurrency(string baseCurrency, string targetCurrency, double amount)
        {
            string apiKey = "fxr_live_43a3c877ea7587d0f2c2593af3df2867736a";
            string apiUrl = $"https://api.fxratesapi.com/latest?apikey={apiKey}&base={baseCurrency}&symbols={targetCurrency}";

            using HttpClient client = new HttpClient();
            HttpResponseMessage response = await client.GetAsync(apiUrl);

            if (!response.IsSuccessStatusCode)
            {
                throw new Exception("Failed to retrieve data from the API.");
            }

            string jsonResponse = await response.Content.ReadAsStringAsync();
            var data = JsonSerializer.Deserialize<ExchangeRateResponse>(jsonResponse, new JsonSerializerOptions 
            {
                PropertyNameCaseInsensitive = true
            });

            if (data != null && data.Rates.TryGetValue(targetCurrency, out double rate))
            {
                return amount * rate;
            }
            else
            {
                throw new Exception("Target currency not found in the API response.");
            }
        }
    }

    public class ExchangeRateResponse
    {
        public bool Success { get; set; }
        public string Terms { get; set; }
        public string Privacy { get; set; }
        public long Timestamp { get; set; }
        public DateTime Date { get; set; }
        public string Base { get; set; }
        public Dictionary<string, double> Rates { get; set; }
    }
}